feature: nh build tool that shows a tree graph of the diffs on rebuild

saylesss88 opened the pull request:
I think nh is a better build tool personally and many agree.

To rebuild use the following command while in the flake directory:

```bash
nh os switch --hostname ghilston .
nh os test --hostname ghilston .
```

You can also use path/to/flake instead of `.`, this is
actually what the `default` value sets in the nh config
so you should be able to omit the path completely and
use `nh os switch`.

I personally have the following shell aliases set for zsh, we can
do the same for you if you like it and want to obv lol

```nix
    shellAliases = {
      sv = "sudo nvim";
      fr = "nh os switch --hostname magic /home/jr/flake";
      ft = "nh os test --hostname magic /home/jr/flake"; # dont save generation to boot menu
      fu = "nh os switch --hostname magic --update /home/jr/flake";
      ncg = "nix-collect-garbage --delete-older-than 3d && sudo nix-collect-garbage -d && sudo /run/current-system/bin/switch-to-configuration boot";
      nc = "nix flake check";
      nt = "nix flake test";
      opts = "man home-configuration.nix";
    }
```

GregHilston reviewed (commented):
Q: Any idea why I get command not found when trying to run `$ nh` after applying this config?

GregHilston commented on nixos/modules/programs/tui/nh.nix:
```diff
@@ -0,0 +1,34 @@
+    };
+    flake = lib.mkOption {
+      type = lib.types.str;
+      default = "${builtins.getEnv "HOME"}/toolbox/nixos";
```
@saylesss88 Something I didn't think about or realize is how we have this filepath hard coded. I imagine this would, or could work on your test machine, but on mine this is not the file path I use.

For example, I use: `~/Git/toolbox` for my repo, and I use `~/Git/saylesss88-toolbox` to test against your branches.

This approach of hard coding the file path does not allow me to easily try out your repo, and then switch to mine.

Do you think we can modify our file here, and our `$ fr` aliases, and other aliases, so that we can pass in a filepath of where our `flake.nix` file lives? So we'd do something like:

`$ fr .` 

or

`$ fr ~/Git/toolbox`

or

`$ fr ~/Git/saylesss88-toolbox`

If this is possible, a nice to have thing would be our alias would default to say `~/Git/toolbox`, if no path is given.

What do you think about this?

saylesss88 commented:
I moved the nh commands to the justfile with a default `flake_path` set to `~/Git/toolbox`.

Usage:

- `just fr mines` will default to building for the `~/Git/toolbox` flake path.

To override this, you can use the following while in the flake directory:

```bash
just fr  mines flake_path=.
```

The setting at the top of the justfile can be adjusted without rebuilding:

```just
set shell := ["zsh", "-cu"]
flake_path := "~/Git/toolbox"
```

So if you didn't want to have to pass the `flake_path=.` you could first change the above setting in your justfile to say `flake_path := "~/Git/saylesss88-toolbox" and you could then rebuild with only `just fr mines`.

For now I left the default to what it was, so if you are working with the repo with your flake directory at `~/toolbox/nixos` you can just run `nh os switch` to rebuild.

I rebased on my end and pushed, that should resolve any conflicts. I'm trying to improve my git workflow to make this a smoother process with a clear linear history.

GregHilston commented:
@saylesss88 

1. Something is broken with our new `Just` approach. We are unable to pass our `flake_path` argument

```
$ just fr  mines flake_path=.
error: Justfile does not contain recipe `flake_path=.`
```

2. Additionally, may we change the bath `flake_path` to be `~/Git/toolbox/nixos`. The `/nixos` is required, as that is where the `flake.nix` lives. Otherwise we'll error out
3. Out of curiosity, what made you add some things to our `Justfile`, and others to our `zsh` alias? Was the former done with commands that take in args, and the latter those that only have hard coded defaults? I'm just curious as to thought process here, I'm not challenging it ;)

GregHilston commented:
@saylesss88 can you comment on 3? 

> Out of curiosity, what made you add some things to our Justfile, and others to our zsh alias? Was the former done with commands that take in args, and the latter those that only have hard coded defaults? I'm just curious as to thought process here, I'm not challenging it ;)
